test/utils.go

```diff
@@ -0,0 +1,133 @@
+// Copyright © 2024 Meroxa, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package testutils
+
+import (
+	"context"
+	"math"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/service/kinesis"
+	"github.com/aws/aws-sdk-go-v2/service/kinesis/types"
+	"github.com/conduitio-labs/conduit-connector-kinesis/common"
+	sdk "github.com/conduitio/conduit-connector-sdk"
+	"github.com/google/uuid"
+	"github.com/matryer/is"
+)
+
+func GetRecords(ctx context.Context, is *is.I, streamName string) []types.Record {
+	client := NewTestClient(ctx, is)
+
+	streamData, err := client.DescribeStream(ctx, &kinesis.DescribeStreamInput{
+		StreamName: &streamName,
+	})
+	is.NoErr(err)
+
+	listShards, err := client.ListShards(ctx, &kinesis.ListShardsInput{
+		StreamARN: streamData.StreamDescription.StreamARN,
+	})
+	is.NoErr(err)
+
+	var recs []types.Record
+	for _, shard := range listShards.Shards {
+		si, err := client.GetShardIterator(ctx, &kinesis.GetShardIteratorInput{
+			ShardId:           shard.ShardId,
+			ShardIteratorType: types.ShardIteratorTypeTrimHorizon,
+			StreamARN:         streamData.StreamDescription.StreamARN,
+		})
+		is.NoErr(err)
+
+		getRecs, err := client.GetRecords(ctx, &kinesis.GetRecordsInput{
+			StreamARN:     streamData.StreamDescription.StreamARN,
+			ShardIterator: si.ShardIterator,
+		})
+		is.NoErr(err)
+
+		recs = append(recs, getRecs.Records...)
+	}
+
+	return recs
+}
+
+func NewTestClient(ctx context.Context, is *is.I) *kinesis.Client {
+	client, err := common.NewClient(ctx, &http.Client{}, common.Config{
+		AWSRegion:          "us-east-1",
+		AWSAccessKeyID:     "accesskeymock",
+		AWSSecretAccessKey: "accesssecretmock",
+		AWSURL:             "http://localhost:4566",
+	})
+	is.NoErr(err)
+
+	return client
+}
+
+// SetupTestStream creates a test stream and returns the name of the stream and a function to delete the stream.
+func SetupTestStream(ctx context.Context, is *is.I) (streamName string, cleanup func()) {
+	client := NewTestClient(ctx, is)
+
+	streamName = "stream_" + strings.ReplaceAll(uuid.NewString()[:8], "-", "")
+	_, err := client.CreateStream(ctx, &kinesis.CreateStreamInput{
+		StreamName: &streamName,
+	})
+	is.NoErr(err)
+
+	for i := 0; i < 10; i++ {
+		time.Sleep(time.Duration(100*math.Pow(2, float64(i))) * time.Millisecond)
```

Comment:
The kinesis golang sdk creates streams asynchronously, and I haven't found any way to subscribe to stream status changes, so doing an exponential backoff is the simplest way I could come up with to wait for the stream to be ready.

Reply:
@alarbada I'm curious why waiting exponentially? i.e.: if we had a specific sleeping time would be sufficient for the stream to be ready>

Reply:
@raulb well it's a common technique in network programming. The assumption is that if the service isn't ready in 1 second, it might be more probable that it is ready in 2 secs than in another sec.

I'm referring to [this](https://en.wikipedia.org/wiki/Exponential_backoff).


I saw it in multiple services to retry failed requests (eg. [openai](https://platform.openai.com/docs/guides/rate-limits), [google](https://cloud.google.com/memorystore/docs/redis/exponential-backoff), etc).

I also used exponential backoff in the pinecone connector. Their vector writes are async, so to assert a write in tests without making too many calls to the pinecone api this technique was useful.

The only drawback that I can think off is that it might make the tests slower, so that's why In the code example you referenced I used 100ms as a starting duration point (100, 200, 400, ... ms).

Reply:
@alarbada got it! thank you for the explanation.